### DISCUSSION TITLE
Integration tests using WebApplicationFactory

### DISCUSSION THREAD

**Discussion by plombaer** [CoreWCF/CoreWCF#1460] [Q&A]
Hi,
I see in the following code that the request is created by "hand".
[https://github.com/CoreWCF/CoreWCF/blob/main/src/CoreWCF.Http/tests/HttpClientTests.cs](url)
Is there a way to programmatically create a SOAP request based on the contract using for instance Message.CreateMessage ?
I tried it but the message is not accepted by the service.
Message message = Message.CreateMessage(MessageVersion.Soap11, action, request);

Thanks

**Comment by mconnew** [comment]
The difficulty you have is your contract can have a method with multiple parameters which need to be serialized. WCF analyzes the contract and creates the correct serializer for each parameter and builds an operation formatter for it. One way to do this would be to create a ChannelFactory for your contract and then pull the IClientMessageFormatter for your operation from the ChannelFactory. One way to do that is to create an IOperationBehavior which in the ApplyClientBehavior method, gets the formatter from `clientOperation.Formatter`. You add this operation behavior by iterating over ChannelFactory.Endpoint.Contract.Operations to find the operation you are interested in, and then add your custom operation behavior to the OperationBehaviors collection on the operation. When you Open the ChannelFactory, it will execute your operation behavior which captures the formatter. Then call IClientMessageFormatter.SerializerRequest passing in the appropriate MessageVersion and the parameters in the order they appear in the method and it will return a Message object. Depending on the binding you are trying to communicate with, you might also have to set the SOAP-Action http header. You'll need to set the content type correctly too.  

Having said all that, usually with this kind of question there's an easier way to solve the problem. So taking a step back, can you explain the problem you are trying to solve and there's probably an easier less convoluted way to achieve it.
